### DISCUSSION TITLE
Declaring node with pydantic

### DISCUSSION THREAD

**Discussion by 0xelight** [neo4j-contrib/neomodel#698] [Ideas]
First of all thank you for your work with this useful package.
Were you familiar with the `pydantic` package? but using this package to declare a node class with neomodel would be huge.
Check https://docs.pydantic.dev/
Especially with validator fields, using python/annotated typing, openapi export, fastapi integration...

What do you think about that?

**Comment by aanastasiou** [comment]
Hi @0xelight 

`neomodel` handles its own validation, through node/relationship properties to ensure that the constraints of the data model it describes are enforced. (For example, see [here](https://neomodel.readthedocs.io/en/latest/properties.html) and [here](https://neomodel.readthedocs.io/en/latest/relationships.html)).

I have not looked into the openapi, fastapi specifications in detail, but we can probably create mappings between the `neomodel` conventions and a given convention if the same entities would have to be used across systems. The validation rules should match but it is always a good idea to validate at the interface to another system.

It has been proposed as a feature for a future release, to at least start adding type hints (as in, plain simple Python type hints), to every interface in the API.

**Comment by mariusconjeaud** [comment]
That would be interesting to look at indeed !

From my perspective of `neomodel` as a user, I've always used it inside of an API, with `FastAPI` as a framework, which indeed includes `pydantic` and it's very helpful there (to validate what comes in/out of the API).

Maybe the upside of having `neomodel` run its own validation though is that we can more easily map the allowed types and validations with the Neo4j allowed types, not sure if we can extend `pydantic` to do that ?

**Reply by hardchor** [reply]
I think it should be possible to implement base nodes and relationships (e.g. how neontology does it: https://github.com/ontolocy/neontology/blob/main/src/neontology/basenode.py), but definitely not a small change.
